predicators/approaches/grammar_search_invention_approach.py

```diff
@@ -454,12 +455,58 @@ def _halving_constant_generator(
 class _SingleFeatureInequalitiesPredicateGrammar(_DataBasedPredicateGrammar):
     """Generates features of the form "0.feature >= c" or "0.feature <= c"."""
 
-    def enumerate(self) -> Iterator[Tuple[Predicate, float]]:
-        # Get ranges of feature values from data.
-        feature_ranges = self._get_feature_ranges()
-        # Edge case: if there are no features at all, return immediately.
-        if not any(r for r in feature_ranges.values()):
-            return
+    def _get_transition_for_inspection(self) -> Iterator[Tuple[State, State]]:
+        max_traj_len = max(
+            len(traj.actions) for traj in self.dataset.trajectories)
+        if max_traj_len < 2:
+            raise ValueError("Can only use transition_changes enumeration " +
+                             "approach when at least one transition has " +
+                             "length greater than 2")
+        # We get transitions backwards thru the trajectory.
+        # We skip the last transition because we already have the goal
+        # predicates, so we don't need to 'inspect' this transition
+        # to invent new predicates that it is trying to achieve.
+        for idx_into_traj in range(-2, (-1 * max_traj_len) - 1, -1):
+            for traj in self.dataset.trajectories:
+                if len(traj.states) < (idx_into_traj * -1) + 1:
+                    continue
+                yield (traj.states[idx_into_traj - 1],
+                       traj.states[idx_into_traj])
+
+    def _get_changing_feats_for_transition(
+            self, transition: Tuple[State, State]) -> Set[Tuple[Type, str]]:
+        """Given a transition, gets all the features that change during the
+        transition."""
+        changing_types_and_feats: Set[Tuple[Type, str]] = set()
+        for obj in transition[0].data:
+            for feat in obj.type.feature_names:
+                if np.sum(
```

Comment:
nope - i just checked in the terminal that the result of `np.subtract` is itself an array

Reply:
oh lol wait something is super weird... don't we want to be doing `state.get` instead of `state.data.get`? I actually have no idea what the latter will do, I'm surprised it doesn't crash...

Reply:
okay I now understand why this doesn't crash I think, but this seems like a big bug, maybe fixing will lead to even bigger improvements. I would like to know why it led to any improvements in the first place